KIPs/kip-dreaft_sdk_common_architecture.md

```diff
@@ -0,0 +1,927 @@
+discussions-to: https://github.com/klaytn/kips/issues/35
+status: Draft
+type: Standards Track
+category: SDK
+created: 2020-07-02
+---
+
+# Table of contents
+
+- [Simple Summary](#simple-summary)
+- [Abstract](#abstract)
+- [Motivation](#motivation)
+- [Specification](#specification)
+  - [Terminology](#terminology)
+  - [Overview of the Common Architecture](#overview-of-the-common-architecture)
+  - [Layer Diagram of the Common Architecture](#layer-diagram-of-the-common-architecture)
+  - [Account Layer Class Diagram](#account-layer-class-diagram)
+  - [Wallet Layer Class Diagram](#wallet-layer-class-diagram)
+  - [Transaction Layer Class Diagram](#transaction-layer-class-diagram)
+  - [RPC Layer Class Diagram](#rpc-layer-class-diagram)
+  - [Contract, ABI, KCT Layer Class Diagram](#contract-abi-kct-layer-class-diagram)
+  - [Utils Layer Class Diagram](#utils-layer-class-diagram)
+- [Example of the source code](#example-of-the-source-code)
+- [Rationale](#rationale)
+- [Backwards Compatibility](#backwards-compatibility)
+- [Test Cases](#test-cases)
+- [Implementation](#implementation)
+- [Copyright](#copyright)
+
+## Simple Summary
+
+A new software architecture for Klaytn development environment which is shared by all Klaytn SDKs (also known as 'caver' series).
+
+## Abstract
+
+The following standard allows Klaytn's SDK, Caver, to implement a common architecture regardless of programming languages.
+
+## Motivation
+
+Klaytn SDKs (caver-js and caver-java) do not share common terminology and software architecture.
+It makes difficult to implement blockchain applications using multiple programming languages.
+For example, if the server program is implemented in Java and the client is implemented in Javascript,
+the developer should learn two different SDKs that have different terminology and software architecture
+even if those programs utilize a single blockchain platform, Klaytn.
+
+The common architecture is designed to unify the terminology used in the SDKs and easily extend to other languages by establishing a language-independent common architecture.
+
+With the common architecture, we want to achieve two goals:
+
+- Developers can easily implement their application using another language if they have implemented in any Klaytn SDK before.
+- An SDK in another programming language can be implemented relatively easily because the software architecture is commonly defined.
+
+## Specification
+
+### Terminology
+
+| Term        | Description |
+| ----------- | ----------- |
+| [Account](https://docs.klaytn.com/klaytn/design/accounts#overview-of-account-state-and-address) | A data structure containing information about a person's balance or a smart contract. Klaytn provides the function to separate the address of the account from the key pair used in the account, so the account stores the [AccountKey] used in the account. |
```

Comment:
```suggestion
| [Account](https://docs.klaytn.com/klaytn/design/accounts#overview-of-account-state-and-address) | A data structure containing information like balance, nonce, the key pair, etc. Note that the key pair of the account can be changed in Klaytn because the key pair and the account address are decoupled. Please refer to [klaytn account] for more detail.  |
```

Reply:
Since it is called a key pair, it seems that everything from the Account data structure to the private key is stored. Doesn't it seem natural to call it a public key of the key pair?

Reply:
@jimni1222 all right. I will use "private key(s)" instead of "key pair". :)